docs/flow-cache-design-doc.md

```diff
@@ -0,0 +1,178 @@
+# Flow Cache Design Doc
+
+## Overview
+
+This design doc outlines the proposed solution for caching the flow object in
+the OAuth2 exchange between the Client, Ory Hydra, and the Consent and Login
+UIs. The flow object contains the state of the authorization request.
+
+## Problem Statement
+
+Currently, the flow object is stored in the database on the Ory Hydra server.
+This approach has several drawbacks:
+
+- Each step of the OAuth2 flow (initialization, consent, login, etc.) requires a
+  database query to retrieve the flow object, and another to update it.
+- Each part of the exchanges supplies different values (login challenge, consent
+  challenge, etc.) to identify the flow object. This means the database table
+  has multiple indices that slow down insertions.
+
+## Proposed Solution
+
+The proposed solution is to store the flow object in client cookies and URLs.
+This way, the flow object is written only once when the flow is completed and
+the final authorization code is generated.
+
+### Requirements
+
+- The flow object must be stored in client cookies and URLs.
+- The flow object must be secure and protect against unauthorized access.
+- The flow object must be persistent, so that the flow can be resumed if the
+  user navigates away from the page or closes the browser.
+- The flow object must be scalable and able to handle a large number of
+  concurrent requests.
+
+### Architecture
+
+The proposed architecture for the flow cache is as follows:
+
+- Store the flow object in an AEAD encrypted cookie.
+- Pass a partial flow around in the URL.
+- Use a secure connection to protect against unauthorized access.
+
+```mermaid
```

Comment:
Please add sequences showing how this works when the user rejects login / consent